### ISSUE TITLE
winnfsd crashing

### ISSUE THREAD

**Issue by spdionis** [winnfsd/winnfsd#29]
I'm using vagrant with the latest winnfsd binary. The latest version still hasn't fixed crashes and hangs.

Usually after half an hour or an hour nfs hangs, even if in the log there is no failed operation. Sometimes it crashes completely, and sometimes the shared folders simply don't show up in the guest machine.

Vagrant 1.8.5 
winnfsd.exe 2.2.0
all mount options default
windows 7 host, ubuntu 14.04 guest

I'm not sure what other info to give you to help debug.

**Comment by marcharding** [comment]
You got this behavior with previous versions as well, right?

Can you please try mounting with tcp instead of udp?

I know that udp sometimes "hangs" and can reproduce this, but still have not figured out whats wrong exactly.

I have not seen any real crashes however, just hangs. You experience real hard crashed, exe crashed, exception thrown, right? Do you have a reliable way to reproduce these hard crashes?

Can you take a look at this comment and see if that also applies to you? https://github.com/winnfsd/winnfsd/issues/13#issuecomment-238407064

I'm considering adding doctor dump so i can collect crash reports, but this would mean that winnfsd cannot be distributed as a single executable any more.

If you willing to help, i can provide you with a build with doctor dump included so i can collect some crash reports.

**Comment by marcharding** [comment]
One more thing 

> and sometimes the shared folders simply don't show up in the guest machine.

Any chance to reproduce that? That's something i never experienced. Did the mount command itself succeed and the folder is just empty?

**Comment by spdionis** [comment]
I can't reliably reproduce the issues. But they happen when the filesystem is used heavily.

In my vm there are a few python crons and services that run often and hit the filesystem a lot to load the python scripts (I see that in the logging). After I deactivated them the vm remained stable a long time. If I let them run I get either a crash (exception) or missing shared folder in 15-30 minutes.

Today i notice that winnfsd.exe takes up more and more memory each time the crons run. Currently at 700 mb after 4 or 5 runs.

This means it's a memory leak, I am willing to help tracking it down of course. Will test mounting with tcp and see if the behaviour is the same.

**Comment by spdionis** [comment]
Adding the mount options in the line below in my vagrantfile (and running `vagrant reload`) did not change the memory behaviour. Note that I'm letting vagrant start winnfsd, not manually. Also I haven't found a way to check precisely if the mount is tcp or udp.

```
    config.vm.synced_folder "../", "/folder", type: "nfs", mount_options: ['rw', 'vers=3', 'tcp', 'nolock']
```

**Comment by marcharding** [comment]
I know about the memory leak, i think it started when the new filetable was implemented (see https://github.com/winnfsd/winnfsd/commit/a506962a6ac635aba15e9ef12f9b1fa2ff89e240). I suspect  your python process creates **a lot** of new files?

As for tcp: That won't change the memory behavior. Your mount options are ok, you can check the actual mount with `mount -v` or `nfsstat -m`.

**Comment by spdionis** [comment]
The python processes are not maintained by me and I don't know what they do in depth, but I suppose they mostly read, instead of creating files. Want me to make sure of it?

**Comment by marcharding** [comment]
hey can you give this version a try [WinNFSd.zip](https://github.com/winnfsd/winnfsd/files/438091/WinNFSd.zip).

Its probably slower and may have some other bugs, since i just rolled back to a older, but much simpler filetable implementation (just to see if thats the main problem).

**Comment by spdionis** [comment]
Unfortunately I couldn't really check for the memory leak because it hanged at the first request on a SETATTR operation.

**Comment by marcharding** [comment]
Hm, that should happen with UDP but not with TCP.

Any chance you can provide me with the most simple test script to reproduce this?

**Comment by spdionis** [comment]
Reproduce what exactly?

**Comment by marcharding** [comment]
The SETATTR hangs.
I know these happen with UDP but never had one with TCP. 
If i could reproduce that, i can probably fix it for both protocols.

**Comment by marcharding** [comment]
@spdionis can you try if [this](https://github.com/winnfsd/winnfsd/files/507074/WinNFSd.zip) version with fixes from @gregordoltar helps with the memory leak?

[WinNFSd.zip](https://github.com/winnfsd/winnfsd/files/507074/WinNFSd.zip)

**Comment by spdionis** [comment]
will do tomorrow :+1:

**Comment by marcharding** [comment]
Any news here ❓

**Comment by spdionis** [comment]
I was sure I left a comment :/ It crashed on startup unfortunately.

**Comment by marcharding** [comment]
That's not good, the tests do finish without problems. 

Seems like some sort of crash reporting needs to be added so we can track this down.

**Comment by marcharding** [comment]
Does version [2.3.0](https://github.com/winnfsd/winnfsd/releases/download/2.3.0/WinNFSd.exe) also crash immediately?

**Comment by spdionis** [comment]
Version 2.3.0 starts fine and seems to work without problems. The memory leak is still present tho.

**Comment by marcharding** [comment]
Thanks! Just wanted to make sure that i did not introduced some bug in 2.3.0.

Will report back with a new version in the next days.

**Comment by gregordoltar** [comment]
@marcharding 

regarding memory leak. I was using http://www.drmemory.org/ to discover memory leaks.
There is however one catch. winnfsd creates a virtual copy of a filesystem in memory and this memory is never freed. This is ok because memory is anyway freed upon program termination. But dr memory will demonstrate it as leak.

**Comment by spdionis** [comment]
> This is ok because memory is anyway freed upon program termination. But dr memory will demonstrate it as leak.

Yes but it's not nice for winnfsd.exe to consume 2gb ram on my machine. Depending on what I do with the vm i get anywhere between 800mb and 2gb ram consumption.

**Comment by gregordoltar** [comment]
@spdionis 

Is memory usage then steady or does it grow ?
How does your export folder look like ? Do you have simple filesystem or do you have links to other folders/files ? When does it start to grow ? For example: 

#1 if you mount your export folder and do "find ." on it. What is the memory footprint ?
#2 if you only do reads what is the footprint then ?
#3 does it happen when you perform writes ?

**Comment by spdionis** [comment]
@gregordoltar 

* `find .` does show memory rising, and after ending the command the memory is not taken back. It does not rise immediately, but only after running for 10 seconds or so. I think it started rising immediately on the first run (want me to retest?). I suppose this is because of some caching mechanism.
* In my tests I cannot say precisely that reads or writes impact memory more or less heavily. I have noticed that when I run my symfony application using the `prod` environment memory usage rarely rises. When i use `dev` environment or delete cache memory usage rises always, and significantly. I thought the issue happened only on writes but this doesn't explain the `find .` behaviour.

So after this I wrote a small script that created a hundred files in a folder inside the export. First time I ran it memory grew, while the second time it didn't. It rised again when I changed the filenames. Running `find` on the folder did not raise memory. After a restart of the vm + winnfsd, running `find` did raise memory.

It looks like there is a file path cache, and adding an element to it causes memory leaks. In the case of symfony dev env/cache a lot of files with random names are generated which are probably added to the cache causing big leaks. (Same for the python scripts i talked before).

**Comment by GM-Alex** [comment]
You guys should think how nfs can speed up the file reads. Yes by caching the files to the ram. So the find command definitely should rase the memory until the files aren't deleted. Only delete should decrease the memory consume.

**Comment by marcharding** [comment]
Files are not cached in ram.

The "virtual filesystem" as @gregordoltar calls it is needed to map nfs file handles (ids) to the the actual file path. On *nix system this is quite easy, one can just use the  inode number. On windows this is not that easy. While ntfs has unique file id (per drive), they do not exist for fat32.

The current implementation leaks memory, because every time a path is requested, the memory the string consumed is not freed, see https://github.com/winnfsd/winnfsd/blob/master/src/FileTree.cpp#L284.

@gregordoltar fixed that and some other smaller leaks, but there seem to be some issues, e.g. @spdionis reports crashes.

That being said it is inevitable for winnfsd to consume more and more ram, due to the "virtual filesystem" layer; but ram usage should not increase when no new files appear nor should it reach 1 GB so fast.

**Comment by spdionis** [comment]
I would close this issue, maybe in favor of some more specific issue. I have not had crashes for a while now with the latest version and memory usage is fine for most use cases. 

Originally I opened the issue with much worse problems that apparently were caused by crons running python, which wrote lots of `.pyc` files trashing the nfs share. This issue doesn't have a goal anymore except a general "improve performance" idea.

Thanks for the support.

**Comment by Gamesh** [comment]
i'm having probably the same issue, after a while winnfsd just hangs, i'm using it with vagrant also. I would say in about 1-2hours of runtime the nfs just hangs and you can't even cd into synced folder on the VM.
winnfsd does not show any CPU usage on windows, the RAM is at 5MB, but i must restart the VM in order to make it operational again

**Comment by Gamesh** [comment]
So i noticed that the hang only happens when using TCP, but if i switch to UDP everything works as expected.

**Comment by Toilal** [comment]
Could you try this exe ? It includes #53 and #51 and fix and seems to fix memory related crashes.

[WinNFSd.zip](https://github.com/winnfsd/winnfsd/files/1351251/WinNFSd.zip)

**Comment by Toilal** [comment]
@marcharding I think issue reported by @Gamesh is not the same as the initial one, maybe you could create another one ?

**Comment by Gamesh** [comment]
this issue was the closes i could find the issues list. Anyway i tried this exe, is still hung after a while, so no change.

**Comment by Toilal** [comment]
It hungs for me too, but fix another crash related to memory leak. I'm now testing with udp, can you confirm the deadlock only occurs with tcp mode?

**Comment by Gamesh** [comment]
yes, it only occurs if set to TCP, if switched to UDP, the hang does not occur.
my other mount options are:
```
rw,noatime,nodiratime,sync,vers=3,rsize=32768,wsize=32768,namlen=255,hard,proto=udp,timeo=11,retrans=3,sec=sys,mountaddr=10.1.10.1,mountvers=3,mountport=1058,mountproto=udp,fsc,local_lock=none,addr=10.1.10.1
```

**Comment by Yasushi** [comment]
https://github.com/winnfsd/winnfsd/blob/4e412a75494a29b71d10b3be2b90460e96e3828c/src/RPCServer.cpp#L107-L109

If set to TCP, WinNFSd will ignore the record fragment header, so the client hangs up.

**Comment by Toilal** [comment]
Great Yasushi. Do you think we only have to remove this condition ? And why does it hung after a while only ? it should hung at first call, isn't it ?

**Comment by Yasushi** [comment]
It is necessary to rebuild the fragment record properly.
The timing to hang is the moment we received a fragment record.
I did not know the condition for the client to send fragment records.

**Comment by marcharding** [comment]
For me this occurs when using udp and not tcp.

**Comment by marcharding** [comment]
The `PORTMAP GETPORT 100024 0` occur because there is no such program, see
https://github.com/winnfsd/winnfsd/blob/5f7f224a3a09a60f61a8b2643f945084952ae172/src/winnfsd.cpp#L23-L25

Wiki says some status daemon should be there, see https://en.wikipedia.org/wiki/Portmap

I'll look into it if i have some time. I'll split that from this issue.

--

I think this issue can be closed?

**Comment by courtney-miles** [comment]
I'm having issues with mount to winnfsd hanging also.

I'm running Winnfsd 2.4.0 with Vagrant 2.0.3 / Virtual Box 5.2.8) via the [winnfsd/vagrant-winnfsd](https://github.com/winnfsd/vagrant-winnfsd) plugin.

To clarify the behaviour, winnfsd does not itself become unresponsive -- I can still run commands like list and reset from it's console. And it does not display any errors when the mount hangs from from my virtual box.

I have tried switching from `tcp` to `udp` -- both have the same problem.

At times, when reloading the vagrant box I do get back an error, but not always -- mostly it just hangs

> `mount.nfs: mount to NFS server '192.168.33.1:/C/Users/Courtney/projects/ZBan' failed: RPC Error: Program unavailable`

or 

> `mount.nfs: Connection timed out`

The frustrating part is whilst sometimes quiting winnfds and then restarting vagrant will get things working again, whilst other times I have to reboot my PC to get them talking again.

If there is currently no fix, I would love to hear if there's a strategy to recover that does not require rebooting my PC each time.

**Comment by bfester** [comment]
Same problem here. 
Host Windows 10 Pro, Guest Ubuntu Xenial, Winnfsd 2.4.0, Vagrant 2.2.1, Virtualbox 5.2.22
Guest machine freezes completely while performing operations on shared files (even reading).
Restarting winnfsd and `vagrant reload` doesn't help. Only full PC reboot does.

**Comment by marcharding** [comment]
While this may not be the most favorable answer, maybe give this project a try (i'm not sure if i can fix whatever is happening):

https://github.com/gfi-centre-ouest/nfs4j-daemon
https://github.com/gfi-centre-ouest/vagrant-nfs4j
